### PR TITLE
Functions for Perfect Number

### PR DESCRIPTION
**Pull request by jhanwar**
#### References to other Issues or PRs

part of #15926 

#### Brief description of what is fixed or changed
Added mersenne_prime_exponent ,is_perfect and is_mersenne_prime
#### Other comments

Already been reviewed  at #16055 
#### Release Notes

- ntheory
     - added ischeck functions for both perfect number and mersenne prime using list of mersenne prime exponents.

### CODE REVIEW

**Comment by jhanwar** [comment]
@smichr @Yathartha22 
These commits has been already reviewed by you on #16055

File: sympy/ntheory/factor_.py [hunk 1]
```diff
@@ -2040,3 +2048,83 @@ def eval(cls, n):
                 raise ValueError("n must be a positive integer")
             else:
                 return sum(factorint(n).values())
+
+
+def mersenne_prime_exponent(nth):
+    """
```
Comment (smichr) [inline sympy/ntheory/factor_.py]: Start Returns right after """ and leave blank line (but no space) before Examples

File: sympy/ntheory/factor_.py [hunk 2]
```diff
@@ -2040,3 +2048,83 @@ def eval(cls, n):
...
+    if n < 1:
+        raise ValueError("nth must be a positive integer; mersenne_prime_exponent(1) == 2")
+    if n > 51:
+        raise ValueError("There are only 51 perfect numbers; nth must be less than or equal to 51")
+    return MERSENNE_PRIME_EXPONENTS[n - 1]
+
+
+def is_perfect(n):
+    """
+    Checks whether n is a perfect number or not. A perfect number is
```
Comment (smichr) [inline sympy/ntheory/factor_.py]: start docstring after """
```
Returns True if ``n`` is a perfect number, else False.

A perfect number is equal to the sum of its positive, proper divisors.
```

File: sympy/ntheory/factor_.py [hunk 3]
```diff
@@ -2040,3 +2048,83 @@ def eval(cls, n):
...
+    n, x = divmod(1 + r, 4)
+    if x:
+        return False
+    e, b = integer_log(n, 2)
+    return b and (e + 1) in MERSENNE_PRIME_EXPONENTS
+
+
+def is_Mersenne_prime(n):
+    """
+    Checks whether n is a Mersenne prime or not.
```
Comment (smichr) [inline sympy/ntheory/factor_.py]: Again, Checks should go after """

**Comment by smichr** [comment]
> About #15926

The keywords that automatically close the Issue are [closes, fixes, resolves](https://help.github.com/en/articles/closing-issues-using-keywords)

**Review by smichr — COMMENTED** [review]
> NO ENTRY

You are adding new functions, so a note that will appear in the release notes is appropriate here.

File: sympy/ntheory/factor_.py [hunk 4]
```diff
@@ -2040,3 +2048,83 @@ def eval(cls, n):
...
+
+    References
+    ==========
+
+    .. [1] http://mathworld.wolfram.com/MersennePrime.html
+
+    """
+    from sympy.core.power import integer_log
+
+    r, b = integer_log(n+1, 2)
```
Comment (Yathartha22) [inline sympy/ntheory/factor_.py]: `n+1` -> `n + 1`

File: sympy/ntheory/factor_.py [hunk 5]
```diff
@@ -2040,3 +2048,83 @@ def eval(cls, n):
...
+    if not b:
+        return False
+    n, x = divmod(1 + r, 4)
+    if x:
+        return False
+    e, b = integer_log(n, 2)
+    return b and (e + 1) in MERSENNE_PRIME_EXPONENTS
+
+
+def is_Mersenne_prime(n):
```
Comment (Yathartha22) [inline sympy/ntheory/factor_.py]: Function name should be in lower-case.

File: sympy/ntheory/factor_.py (hunk 1 above)
Comment (jhanwar) [reply sympy/ntheory/factor_.py]: Done

File: sympy/ntheory/factor_.py (hunk 2 above)
Comment (jhanwar) [reply sympy/ntheory/factor_.py]: Done

File: sympy/ntheory/factor_.py (hunk 3 above)
Comment (jhanwar) [reply sympy/ntheory/factor_.py]: Done

File: sympy/ntheory/factor_.py (hunk 4 above)
Comment (jhanwar) [reply sympy/ntheory/factor_.py]: Done

File: sympy/ntheory/factor_.py (hunk 5 above)
Comment (jhanwar) [reply sympy/ntheory/factor_.py]: Done

**Comment by smichr** [comment]
Looks good. And now I see why you put "addresses" in front of the issue number in the OP: because it doesn't close the issue, only addresses part of it. I will update the issue.

**Comment by Yathartha22** [comment]
LGTM. +1

**Pull request merged**